File: portfolio-backend/services/jobs-resume/services/sunset.py
"""AI feature sunset — Resume Tailor moved to Aspirely (https://aspirely.me).

The portfolio's AI resume tooling is retired. Bedrock InvokeModel has been
revoked from the Lambda execution role, so any call that would have reached a
model now fails at the IAM layer with an opaque AccessDenied. This module
turns that into one deliberate, honest answer everywhere.

Two layers, on purpose:

  1. **Route guard** (`sunset_guard`, wired as a `before_request` on the jobs
     and resume blueprints). Retired endpoints answer 410 immediately — no
     Lambda fan-out, no async job row, no polling loop that ends in a generic
     "failed". This is what the UI actually renders.

  2. **Provider guard** (`llm_providers.claude` raises `AIFeatureSunsetError`).
     Backstop for any AI path not named in `RETIRED_ENDPOINTS`. The app-level
     error handler renders it with the same payload as layer 1, so a missed
     route degrades to the correct message rather than a 500.

What deliberately still works: job discovery (Workday, Career Pages, daily
pipeline, saved jobs, match scores — all deterministic, no model calls), and
full read/download access to every resume and tailoring record already stored.
Nobody loses data; they just can't generate new AI output here.
"""
from flask import jsonify

NEW_PRODUCT_NAME = "Aspirely"
NEW_PRODUCT_URL = "https://aspirely.me"

#: Machine-readable discriminator. The frontend switches on this to render the
#: migration card instead of a generic error toast.
SUNSET_CODE = "ai_features_moved"

#: One sentence, shown as the card headline.
SUNSET_TITLE = "AI resume tailoring has moved to Aspirely"

#: The body. Says what stopped, what didn't, and where to go — in that order,
#: because "did I lose my data?" is the first thing anyone will wonder.
SUNSET_MESSAGE = (
    "This portfolio no longer tailors resumes. Your resumes, tailored versions, "
    "and application history are all still here and still downloadable — nothing "
    "was deleted. Job search also still works. "
    f"Tailoring, cover letters, ATS scoring, interview prep and auto-apply now live "
    f"on {NEW_PRODUCT_NAME} — {NEW_PRODUCT_URL}"
)

#: Rendered as a checklist on the migration card. Sourced from the live
#: Aspirely nav (jobs / resume / apply / grow) so it doesn't overpromise.
NEW_PRODUCT_HIGHLIGHTS = [
    "Matched jobs from employer career sites, with fit reasons and sponsorship signals",
    "Resume tailoring and batch tailoring against many roles at once",
    "Auto-Apply — queue roles and let Aspirely fill the application forms",
    "Application tracker, managed inbox, and interview prep",
    "Career Copilot, learning sprints, and referral network tools",
]


class AIFeatureSunsetError(Exception):
    """Raised when code reaches an LLM call that no longer exists.

    Intentionally NOT a subclass of `LLMRetriesExhaustedError`. Several
    callers (match_gap_reporter, filter_suggester) catch that type and quietly
    fall back to a degraded-but-plausible result — correct behaviour for a
    transient model outage, wrong here, where the honest answer is "this
    feature is gone, go to Aspirely". Keeping the types separate means the
    error propagates to the handler instead of being swallowed.
    """

    code = SUNSET_CODE


def sunset_payload() -> dict:
    """The response body shared by the route guard and the error handler."""
    return {
        "error": SUNSET_CODE,
        "code": SUNSET_CODE,
        "title": SUNSET_TITLE,
        "message": SUNSET_MESSAGE,
        "moved_to": NEW_PRODUCT_URL,
        "product": NEW_PRODUCT_NAME,
        "highlights": NEW_PRODUCT_HIGHLIGHTS,
        "data_retained": True,
        "still_available": [
            "Job search (Workday, Career Pages, daily pipeline)",
            "Downloading resumes and tailored versions you already have",
            "Application tracker history",
        ],
    }


def sunset_response():
    """410 Gone + the migration payload.

    410 rather than 404 (the route existed) or 503 (nothing is coming back
    up). Clients that only look at status get a correct signal; clients that
    read the body get the full story.
    """
    return jsonify(sunset_payload()), 410


# ---------------------------------------------------------------------------
# Retired endpoints
# ---------------------------------------------------------------------------
# Flask endpoint names ("<blueprint>.<view_func>"). Anything listed here used
# to call a model to GENERATE something. Read, list, download, delete and
# job-discovery endpoints are deliberately absent — they still work.
RETIRED_ENDPOINTS = frozenset({
    # ── Tailoring core ────────────────────────────────────────────────────
    "resume.extract_jd",
    "resume.tailor",
    "resume.tailor_with_jd",
    "resume.regenerate",
    "resume.rewrite_bullet",
    "resume.ats_scores",
    "resume.cover_letter",
    "resume.cover_letter_download",
    "resume.batch_tailor",
    # Uploading a NEW base resume runs the LLM parser, so it goes too.
    # Existing parsed resumes are untouched and still served by /status,
    # /versions, /generated and /download-file.
    "resume.upload",
    # ── Interview prep (all generative) ───────────────────────────────────
    "resume.generate_interview_prep",
    "resume.practice_question",
    "resume.interview_chat",
    "resume.mock_evaluate",
    # ── Career Copilot — generative surfaces only. The CRUD underneath
    #    (contacts, offers, memory, momentum, timeline) stays readable.
    "resume.career_copilot_chat",
    "resume.career_copilot_behavior",
    "resume.career_copilot_playground_start",
    "resume.career_copilot_playground_advance",
    "resume.career_copilot_playground_quiz",
    "resume.career_copilot_playground_evaluate",
    "resume.career_copilot_outreach_generate_sequence",
    "resume.career_copilot_intelligence_followup",
    "resume.career_copilot_intelligence_reframe",
    "resume.career_copilot_intelligence_rejection_insights",
    "resume.career_copilot_intelligence_weekly_digest",
    "resume.career_copilot_network_generate_intro",
    "resume.career_copilot_network_insights",
    "resume.career_copilot_offers_negotiate",
    # ── Beta lab (model-backed experiments) ───────────────────────────────
    "resume.beta_morning_brief",
    "resume.beta_similar_roles",
    # ── Gmail classification pass ─────────────────────────────────────────
    # Linking, listing and dismissing suggestions still work; only the sync
    # that runs messages through a model is retired.
    "resume.gmail_sync",
    # ── Jobs blueprint — generative helpers. Search / pipeline / catalogs
    #    are deterministic and stay live.
    "jobs.analyze",
    "jobs.tailor_resume",
    "jobs.draft_outreach_route",
    "jobs.suggest_pipeline_filters",
    "jobs.upload_resume",
    # ── Apply (Chrome extension answer generation) ────────────────────────
    "apply.generate_answer",
})


def sunset_guard():
    """`before_request` hook: 410 on retired endpoints, pass through otherwise.

    Returning None lets the request continue, which is what every kept
    endpoint does. CORS preflight is never blocked — OPTIONS has to succeed or
    the browser reports a CORS failure and the UI can't read our 410 body.
    """
    from flask import request

    if request.method == "OPTIONS":
        return None
    if request.endpoint in RETIRED_ENDPOINTS:
        return sunset_response()
    return None

File: portfolio-frontend/src/lib/sunset.ts
/**
 * Resume Tailor sunset — AI features moved to Aspirely (https://aspirely.me).
 *
 * One source of truth for the migration copy and for WHICH parts of the
 * workspace are retired. `ResumeParser.tsx` derives its nav badges and its tab
 * bodies from `RETIRED_TABS` below, so a tab can never be marked "Moved" in
 * the sidebar while still rendering a live-looking form.
 *
 * The split mirrors the backend's `services/sunset.py`: anything that used to
 * call a model is retired; job discovery and read/download of existing data
 * are not. Keep the two lists in agreement — the backend answers 410 either
 * way, but a user reaching a dead form is a worse experience than never being
 * offered it.
 */

export const ASPIRELY_NAME = "Aspirely";
export const ASPIRELY_URL = "https://aspirely.me";

/** Headline for the full-page migration card. */
export const SUNSET_TITLE = "Resume tailoring has moved to Aspirely";

/**
 * The short version, reused by toasts and the API error path. Leads with the
 * reassurance — "did I lose my work?" is the first question anyone has.
 */
export const SUNSET_MESSAGE =
  "This tool no longer tailors resumes. Your resumes and history are still here " +
  `and still downloadable — nothing was deleted. Tailoring now lives on ${ASPIRELY_NAME} (${ASPIRELY_URL}).`;

/** Still functional here — stated plainly so nobody assumes the whole app died. */
export const STILL_WORKS: readonly string[] = [
  "Job search — Workday, Career Pages, and the daily pipeline",
  "Downloading every resume and tailored version you already made",
  "Your applications, saved jobs, and visa timeline",
];

/**
 * What Aspirely does, taken from its live navigation rather than a pitch deck,
 * so the list stays honest about what someone will actually find there.
 */
export const ASPIRELY_HIGHLIGHTS: readonly { title: string; detail: string }[] = [
  {
    title: "Jobs matched to your resume",
    detail:
      "Roles pulled straight from employer career sites, scored against your resume, with fit reasons and sponsorship signals.",
  },
  {
    title: "Tailoring and batch tailoring",
    detail:
      "Rewrite one resume for one job description, or queue many roles at once. Cover letters and ATS scoring included.",
  },
  {
    title: "Auto-Apply",
    detail:
      "Queue the roles you want and let Aspirely fill in the application forms, with as much or as little approval as you set.",
  },
  {
    title: "Tracker, inbox, and interview prep",
    detail:
      "Every application and where it stands, replies to a managed application address, and practice questions with a mock interviewer.",
  },
  {
    title: "Career Copilot and Learning Hub",
    detail:
      "Ask about your search and get a concrete next step. Close skill gaps with a guided sprint, and find referrals worth chasing.",
  },
];

/**
 * Workspace tabs whose entire purpose was model-generated output. These render
 * the migration card instead of their old body.
 *
 * Not retired, on purpose: `my-resumes`, `tailored`, `applications`, `jobs`,
 * `workday-jobs`, `career-pages`, `visa`, `profile` — every one of those is
 * either job discovery or reading back data the user already owns.
 */
export const RETIRED_TABS: readonly string[] = [
  "tailor",
  "batch",
  "interview",
  "copilot",
  "beta",
];

export function isRetiredTab(tab: string): boolean {
  return RETIRED_TABS.includes(tab);
}
